features/rbac: TestManager_CreateStewardSubject TempDir cleanup race blocks CI across PRs

## Goal

`features/rbac` tests that use the `flatfile` + `sqlite` OSS storage stack shut down cleanly on every platform — no `TempDir RemoveAll cleanup: directory not empty` failures, no race between `t.TempDir()` cleanup and a still-running audit drain goroutine. The fix addresses the production shutdown path (not just the two tests currently known to flake) so the entire `features/rbac` test suite and any future test using the same fixture benefits.

**Desired end state:**
- Running `go test ./features/rbac/... -count=100` on macos-latest produces zero `TempDir RemoveAll cleanup` failures. (macos-latest is the primary recurrence platform; must also pass on Linux and Windows.)
- Running the suite in the GitHub merge queue context stops dequeuing PRs due to this race.
- The root cause — a background audit drain goroutine (`pkg/audit.Manager.drainLoop`) outliving the test because nothing shuts it down — is fixed in production code by ensuring the audit manager has a synchronous shutdown path and the RBAC manager wires through to it.
- All `features/rbac` tests that construct a `storageManager` via `interfaces.CreateOSSStorageManager` register cleanup in the correct order: shut down the RBAC/audit manager first (drains and stops the goroutine), then close the storage stores, then let `t.TempDir()`'s built-in cleanup run.
- Zero `t.Skip` and zero quarantined tests. The fix covers **both** known-flaking tests (`TestManager_CreateStewardSubject`, `TestManager_SystemAdminPermissions`) and any sibling test in the same package that shares the fixture.

## Context

### Observed failures

Two tests have been observed flaking with the same fingerprint, both on `macos-latest`:

1. **`TestManager_CreateStewardSubject`** (original report) — blocked at least PRs #839, #843, #844, #845, #847.
2. **`TestManager_SystemAdminPermissions`** (new, 2026-04-24) — dequeued PR #852 from the merge queue (attempt 3, 13:31→13:38 UTC). Failure captured at https://github.com/cfg-is/cfgms/actions/runs/24892179475/job/72887144117:

```
--- FAIL: TestManager_SystemAdminPermissions (0.24s)
    testing.go:1369: TempDir RemoveAll cleanup: unlinkat /var/folders/tb/.../T/TestManager_SystemAdminPermissions480422175/001/flatfile: directory not empty
FAIL	github.com/cfgis/cfgms/features/rbac	6.194s
```

Same package, same `flatfile` subdirectory residue, same platform bias. Almost certainly the same bug.

### Root cause (verified by code inspection)

`pkg/audit/manager.go` L180-L225: `NewManager` starts a background `drainLoop` goroutine (`go m.drainLoop()`) that consumes from a bounded queue and writes atomic-rename temp files (`.tmp-cfg-*`) into the `flatfile` root.

`features/rbac/manager.go` L73: `NewManagerWithStorage` creates an `audit.NewManager(...)` internally and stores it on the `Manager` struct as `auditManager`. There is no exported shutdown method on the RBAC `Manager` — callers have no way to stop the drain goroutine.

`features/rbac/manager_test.go` (and `manager_storage_test.go`, `manager_hierarchy_test.go`, `audit_integration_test.go` — all share the fixture) registers:

```
tmpDir := t.TempDir()
storageManager, err := interfaces.CreateOSSStorageManager(tmpDir+"/flatfile", tmpDir+"/cfgms.db")
require.NoError(t, err)
t.Cleanup(func() { _ = storageManager.Close() })
manager := NewManagerWithStorage(...)
```

At teardown, Go runs cleanups in LIFO order: `storageManager.Close()` first, then `TempDir`'s `RemoveAll`. But nobody closes the audit manager. Its `drainLoop` is still flushing queued events, calling the flatfile writer which creates a `.tmp-cfg-*` file in `tmpDir/flatfile` and renames it. On slower filesystems (macOS runners in GitHub Actions) this races with `RemoveAll` — a temp file is created between `RemoveAll`'s directory enumeration and unlink, so the directory isn't empty when `RemoveAll` tries to remove it.

The `flatfile` provider's `Close()` methods are no-ops (`FlatFileAuditStore.Close() returns nil` at `pkg/storage/providers/flatfile/audit_store.go:682`) — they can't stop the goroutine because the goroutine lives in `pkg/audit`, not in the storage provider.

### Why the #12775378 pattern applies

Like the #366 fix, the right solution is to expose a synchronous completion signal on the production type (audit manager shutdown), not to paper over the race with sleeps in the test. This story should follow the same precedent.

## Files In Scope

Primary production edits:
- `pkg/audit/manager.go` — confirm / expose a `Close()` (or `Shutdown(ctx)`) method that closes the `stop` channel and blocks on `done` to guarantee the drain goroutine has returned before `Close` returns. It likely already exists internally via `stop`/`done` plumbing — verify, export if needed, add a godoc comment explicitly stating callers MUST call it before disposing of the backing store or temp directory.
- `features/rbac/manager.go` — add a `Close()` (or `Shutdown(ctx)`) method on the RBAC `Manager` that calls the audit manager's shutdown. Godoc should state this must be called before closing the underlying storage.

Primary test edits:
- `features/rbac/manager_test.go` — for every test that creates a `storageManager`, add `t.Cleanup(func() { _ = manager.Close() })` before the `t.Cleanup(storageManager.Close)` call so cleanup runs in the correct LIFO order: audit drain stop → store close → temp dir removal.
- `features/rbac/manager_storage_test.go` — same treatment.
- `features/rbac/manager_hierarchy_test.go` — same treatment.
- `features/rbac/audit_integration_test.go` — same treatment.
- Any other file in `features/rbac/` matching `grep -l "CreateOSSStorageManager" features/rbac/*_test.go` — enumerate and fix all call sites for consistency.

Consider extracting the fixture into a shared helper (e.g. `newTestManager(t)` returning `(*Manager, func())` or registering cleanup internally) to prevent this bug from being re-introduced as new tests are added. Only do this if it doesn't balloon scope — the minimum fix is per-test cleanup registration.

Out of scope:
- Tests in packages other than `features/rbac/` (file a follow-up if the same pattern exists elsewhere; do not broaden this story).
- Changes to the `flatfile` or `sqlite` provider internals (they are correct — the bug is that nobody waits for the audit drain).
- Changes to `pkg/storage/interfaces.StorageManager.Close()` — it already does the right thing for stores.
- Any change to central providers beyond `pkg/audit` (which is a direct provider, not pluggable).

## Dependencies

None. Independent of issue #366 (which is dispatched as `cfg-agent-366` and touches `features/tenant/security/*`). File sets are disjoint.

Sibling coordination note: #366 could in principle touch `pkg/storage/providers/flatfile` if its fix route requires it, but the expected fix for #366 is isolated to `features/tenant/security`. If #366's eventual PR touches `pkg/audit` or the flatfile provider, the #848 agent must rebase onto #366's merged changes before opening a PR — standard merge-queue hygiene, not a hard dependency.

## Acceptance Criteria

- [ ] `TestManager_CreateStewardSubject` passes 100 consecutive runs on Linux with and without `-race`.
- [ ] `TestManager_SystemAdminPermissions` passes 100 consecutive runs on Linux with and without `-race`.
- [ ] `go test ./features/rbac/... -count=50` passes on Linux with no `TempDir RemoveAll cleanup` failures.
- [ ] CI passes on Linux, Windows, and macos-latest (macos-latest is the critical signal — this is where the race manifests).
- [ ] PR passes through the GitHub merge queue without dequeue due to this race (verify post-merge-queue-attempt, not just local runs).
- [ ] `pkg/audit.Manager` has a documented, exported shutdown method with a godoc comment stating the goroutine-cleanup contract. If the method already exists, the comment is updated to make the contract explicit.
- [ ] `features/rbac.Manager` has a documented shutdown method that chains to the audit manager's shutdown. Godoc states the expected call order relative to storage close.
- [ ] Every test file in `features/rbac/` that calls `interfaces.CreateOSSStorageManager` registers cleanup in the correct LIFO order (RBAC manager shutdown → storage close → temp dir). Enumerate them in the PR description to confirm full coverage.
- [ ] No mocks introduced. Real CFGMS components only.
- [ ] No `t.Skip` added anywhere in `features/rbac/`.
- [ ] `make test-complete` passes locally.

## Constraints

- No mocks. Real `audit.Manager`, real `flatfile` + `sqlite` stores (CLAUDE.md zero-tolerance rule).
- No `t.Skip` on any `features/rbac` test. No quarantine.
- Prefer production-code fixes (expose a proper shutdown API on the audit manager and RBAC manager) over test-level polling. Test-only workarounds (e.g. `require.Eventually` waiting for an empty directory) are acceptable ONLY if the production shutdown API genuinely has no synchronous completion signal AND that limitation is called out in the PR description with a follow-up issue filed.
- Deterministic synchronization. No `time.Sleep` based waits for async state.
- Branch from `develop`: `feature/story-848-rbac-tempdir-cleanup-race`.
- PR targets `develop` (never `main`).
- Commit format: `<scope>: <what changed> (Issue #848)` with `Fixes #848` in body.
- No changes to `CLAUDE.md`, `Makefile` root targets, `.github/workflows/`, or central provider interfaces.
- No new Go module dependencies.

## Additional observation — same race affects `TestManager_SystemAdminPermissions`

**New occurrence (2026-04-24, PR #852 merge queue attempt 3):**

- Platform: macos-latest
- Queue branch: `gh-readonly-queue/develop/pr-852-2c8a4681...`
- Job: https://github.com/cfg-is/cfgms/actions/runs/24892179475/job/72887144117

**Failure (same fingerprint as original #848, different test function):**

```
--- FAIL: TestManager_SystemAdminPermissions (0.24s)
    testing.go:1369: TempDir RemoveAll cleanup: unlinkat /var/folders/tb/.../T/TestManager_SystemAdminPermissions480422175/001/flatfile: directory not empty
FAIL	github.com/cfgis/cfgms/features/rbac	6.194s
```

**Implications:**
- Same package (`features/rbac`), same cleanup pattern (`TempDir RemoveAll cleanup: directory not empty`), same `flatfile` subdirectory residue, same macos-latest bias.
- This is almost certainly the same root cause as the original `TestManager_CreateStewardSubject` observation — a storage-layer fd or goroutine outliving the test function.
- Both tests presumably instantiate the same storage stack in setup; a single fix to the flatfile shutdown ordering should cover both (and likely any other `features/rbac` test that sets up the same fixture).

**Scope update suggested:** broaden the bug to "all `features/rbac` manager tests using flatfile storage" rather than a single test name. Acceptance criteria should require 100x runs for *both* `TestManager_CreateStewardSubject` and `TestManager_SystemAdminPermissions`, and ideally a `grep`-style audit of the package to enumerate any sibling tests that share the setup.

**Impact update:** PR #852 was dequeued by this failure (13:31→13:38 UTC today), adding to the list of PRs disrupted by this race.

--- a/features/controller/controller_test.go
+++ b/features/controller/controller_test.go
@@ -48,8 +48,12 @@ func TestControllerCreation(t *testing.T) {
 				if tt.cfg.Certificate != nil {
 					tt.cfg.Certificate.CAPath = tempDir + "/certs/ca"
 				}
-				if tt.cfg.Storage != nil && tt.cfg.Storage.Config != nil {
-					tt.cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+				if tt.cfg.Storage != nil {
+					tt.cfg.Storage.FlatfileRoot = tempDir + "/flatfile"
+					tt.cfg.Storage.SQLitePath = tempDir + "/cfgms.db"
+					if tt.cfg.Storage.Config != nil {
+						tt.cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+					}
 				}
 				// Pre-initialize if cert management is enabled (Story #410)
 				if tt.cfg.Certificate != nil && tt.cfg.Certificate.EnableCertManagement {
@@ -85,8 +89,12 @@ func TestControllerLifecycle(t *testing.T) {
 	if cfg.Certificate != nil {
 		cfg.Certificate.CAPath = tempDir + "/certs/ca"
 	}
-	if cfg.Storage != nil && cfg.Storage.Config != nil {
-		cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+	if cfg.Storage != nil {
+		cfg.Storage.FlatfileRoot = tempDir + "/flatfile"
+		cfg.Storage.SQLitePath = tempDir + "/cfgms.db"
+		if cfg.Storage.Config != nil {
+			cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+		}
 	}
 
 	// Pre-initialize (Story #410: controller requires explicit init)
@@ -167,8 +175,12 @@ func TestModuleRegistration(t *testing.T) {
 	if cfg.Certificate != nil {
 		cfg.Certificate.CAPath = tempDir + "/certs/ca"
 	}
-	if cfg.Storage != nil && cfg.Storage.Config != nil {
-		cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+	if cfg.Storage != nil {
+		cfg.Storage.FlatfileRoot = tempDir + "/flatfile"
+		cfg.Storage.SQLitePath = tempDir + "/cfgms.db"
+		if cfg.Storage.Config != nil {
+			cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+		}
 	}
 	pkgtestutil.PreInitControllerForTest(t, cfg.CertPath, cfg.Certificate.CAPath)
 	ctrl, err := New(cfg, logger)
@@ -219,8 +231,12 @@ func TestControllerSingleHTTPServer(t *testing.T) {
 	if cfg.Certificate != nil {
 		cfg.Certificate.CAPath = tempDir + "/certs/ca"
 	}
-	if cfg.Storage != nil && cfg.Storage.Config != nil {
-		cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+	if cfg.Storage != nil {
+		cfg.Storage.FlatfileRoot = tempDir + "/flatfile"
+		cfg.Storage.SQLitePath = tempDir + "/cfgms.db"
+		if cfg.Storage.Config != nil {
+			cfg.Storage.Config["repository_path"] = tempDir + "/storage"
+		}
 	}
 	if cfg.Transport != nil {
 		cfg.Transport.ListenAddr = "127.0.0.1:0"

--- a/features/controller/server/server_security_test.go
+++ b/features/controller/server/server_security_test.go
@@ -37,8 +37,12 @@ func createTestStorageConfig(tempDir, suffix string) *config.StorageConfig {
 	}
 }
 
-// createDockerTestStorageConfig creates storage configs for Docker-based testing
-func createDockerTestStorageConfig(provider string) *config.StorageConfig {
+// createDockerTestStorageConfig creates storage configs for Docker-based testing.
+// For non-database providers, the flatfile+sqlite paths are scoped to t.TempDir()
+// so each test gets a fresh directory and no state persists in /tmp across runs —
+// stale on-disk schemas had previously broken tests after audit_entries migrations.
+func createDockerTestStorageConfig(t *testing.T, provider string) *config.StorageConfig {
+	t.Helper()
 	switch provider {
 	case "database":
 		return &config.StorageConfig{
@@ -53,7 +57,7 @@ func createDockerTestStorageConfig(provider string) *config.StorageConfig {
 			},
 		}
 	default:
-		return createTestStorageConfig(os.TempDir(), provider)
+		return createTestStorageConfig(t.TempDir(), provider)
 	}
 }
 
@@ -221,7 +225,7 @@ func TestServer_StorageProviderValidation(t *testing.T) {
 
 				// Use Docker test configuration if available, otherwise fall back to local test
 				if isDockerTestEnvironment() {
-					storageConfig = createDockerTestStorageConfig(providerInfo.Name)
+					storageConfig = createDockerTestStorageConfig(t, providerInfo.Name)
 					t.Logf("Using Docker test configuration for provider '%s'", providerInfo.Name)
 				} else {
 					// For local testing, use appropriate configuration per provider

--- a/features/rbac/advanced_permissions_test.go
+++ b/features/rbac/advanced_permissions_test.go
@@ -41,11 +41,7 @@ func TestAdvancedPermissionManagement(t *testing.T) {
 	err = manager.Initialize(ctx)
 	require.NoError(t, err)
 
-	t.Cleanup(func() {
-		stopCtx, stopCancel := context.WithTimeout(context.Background(), 5*time.Second)
-		defer stopCancel()
-		_ = manager.auditManager.Stop(stopCtx)
-	})
+	// Drain shutdown is handled by manager.Close cleanup registered above (Issue #848).
 
 	flushAudit := func(t *testing.T) {
 		t.Helper()

--- a/features/rbac/advanced_permissions_test.go
+++ b/features/rbac/advanced_permissions_test.go
@@ -31,6 +31,11 @@ func TestAdvancedPermissionManagement(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)

--- a/features/rbac/audit_integration_test.go
+++ b/features/rbac/audit_integration_test.go
@@ -33,6 +33,11 @@ func TestRBACManager_AuditIntegration(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	require.NotNil(t, manager)
 	require.NotNil(t, manager.auditManager)
 
@@ -332,6 +337,11 @@ func TestRBACManager_AuditFailureHandling(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	require.NotNil(t, manager)
 
 	ctx := context.Background()
@@ -377,6 +387,11 @@ func TestRBACManager_AuditFailureHandling(t *testing.T) {
 			storageManager.GetClientTenantStore(),
 			storageManager.GetRBACStore(),
 		)
+		t.Cleanup(func() {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			_ = manager.Close(ctx)
+		})
 		err = manager.Initialize(ctx)
 		require.NoError(t, err)
 

--- a/features/rbac/audit_integration_test.go
+++ b/features/rbac/audit_integration_test.go
@@ -46,13 +46,8 @@ func TestRBACManager_AuditIntegration(t *testing.T) {
 	require.NoError(t, err)
 
 	// Issue #764: audit writes are now asynchronous. Tests that query the audit
-	// store must Flush first to guarantee pending entries have landed.
-	t.Cleanup(func() {
-		stopCtx, stopCancel := context.WithTimeout(context.Background(), 5*time.Second)
-		defer stopCancel()
-		_ = manager.auditManager.Stop(stopCtx)
-	})
-
+	// store must Flush first to guarantee pending entries have landed. Drain
+	// shutdown is handled by manager.Close cleanup registered above (Issue #848).
 	flushAudit := func(t *testing.T) {
 		t.Helper()
 		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
@@ -266,37 +261,39 @@ func TestRBACManager_AuditIntegration(t *testing.T) {
 	})
 
 	t.Run("Failed operations generate error audit events", func(t *testing.T) {
-		// Try to create a role with invalid data (empty ID)
+		// Use a case that deterministically fails at the manager level:
+		// a role claiming a non-existent parent triggers the explicit error
+		// path in Manager.CreateRole (RBAC_PARENT_ROLE_NOT_FOUND), which
+		// synchronously emits an error audit event.
 		invalidRole := &common.Role{
-			Id:       "", // Invalid - empty ID
-			Name:     "Invalid Role",
-			TenantId: "test-tenant",
+			Id:           "test-role-parent-missing",
+			Name:         "Role with missing parent",
+			TenantId:     "test-tenant",
+			ParentRoleId: "does-not-exist",
 		}
 
-		_ = manager.CreateRole(ctx, invalidRole)
-		// This should fail, but let's check if we still get an audit event
+		err := manager.CreateRole(ctx, invalidRole)
+		require.Error(t, err, "CreateRole must return an error when parent role does not exist")
+		require.Contains(t, err.Error(), "not found")
 
-		// Query for any audit entries with error result
 		auditFilter := &business.AuditFilter{
-			TenantID:   "test-tenant",
-			EventTypes: []business.AuditEventType{business.AuditEventUserManagement},
-			Actions:    []string{"create_role"},
-			Results:    []business.AuditResult{business.AuditResultError},
-			Limit:      10,
+			TenantID:      "test-tenant",
+			EventTypes:    []business.AuditEventType{business.AuditEventUserManagement},
+			Actions:       []string{"create_role"},
+			Results:       []business.AuditResult{business.AuditResultError},
+			ResourceIDs:   []string{"test-role-parent-missing"},
+			Limit:         10,
 		}
 
-		// Note: This test depends on whether the underlying store validates the role
-		// If validation passes through, we won't get an error audit event
 		flushAudit(t)
 		auditEntries, err := manager.auditManager.QueryEntries(ctx, auditFilter)
 		require.NoError(t, err)
+		require.Len(t, auditEntries, 1, "failed CreateRole must produce exactly one error audit entry")
 
-		// If there are error entries, verify they have proper error information
-		for _, entry := range auditEntries {
-			assert.Equal(t, business.AuditResultError, entry.Result)
-			assert.NotEmpty(t, entry.ErrorCode)
-			assert.NotEmpty(t, entry.ErrorMessage)
-		}
+		entry := auditEntries[0]
+		assert.Equal(t, business.AuditResultError, entry.Result)
+		assert.Equal(t, "RBAC_PARENT_ROLE_NOT_FOUND", entry.ErrorCode)
+		assert.NotEmpty(t, entry.ErrorMessage)
 	})
 
 	t.Run("Audit entries maintain integrity", func(t *testing.T) {

--- a/features/rbac/audit_integration_test.go
+++ b/features/rbac/audit_integration_test.go
@@ -277,12 +277,12 @@ func TestRBACManager_AuditIntegration(t *testing.T) {
 		require.Contains(t, err.Error(), "not found")
 
 		auditFilter := &business.AuditFilter{
-			TenantID:      "test-tenant",
-			EventTypes:    []business.AuditEventType{business.AuditEventUserManagement},
-			Actions:       []string{"create_role"},
-			Results:       []business.AuditResult{business.AuditResultError},
-			ResourceIDs:   []string{"test-role-parent-missing"},
-			Limit:         10,
+			TenantID:    "test-tenant",
+			EventTypes:  []business.AuditEventType{business.AuditEventUserManagement},
+			Actions:     []string{"create_role"},
+			Results:     []business.AuditResult{business.AuditResultError},
+			ResourceIDs: []string{"test-role-parent-missing"},
+			Limit:       10,
 		}
 
 		flushAudit(t)

--- a/features/rbac/manager.go
+++ b/features/rbac/manager.go
@@ -109,6 +109,23 @@ func NewManagerWithStorage(auditStore business.AuditStore, clientTenantStore bus
 	return manager
 }
 
+// Close flushes pending audit entries and stops the audit drain goroutine.
+// It is idempotent via the underlying audit.Manager's stopOnce.
+//
+// Callers that construct a Manager and back it with per-test temporary storage
+// (e.g., via `t.TempDir()`) MUST register a t.Cleanup that calls Close before
+// the temp directory is removed. Without this, the audit drain goroutine may
+// still be writing files when `TempDir` RemoveAll runs, producing flaky
+// "directory not empty" cleanup errors on slower filesystems (macOS, Windows).
+//
+// Returns nil if no audit manager was configured.
+func (m *Manager) Close(ctx context.Context) error {
+	if m.auditManager == nil {
+		return nil
+	}
+	return m.auditManager.Stop(ctx)
+}
+
 // Initialize sets up the RBAC system with default roles and permissions
 func (m *Manager) Initialize(ctx context.Context) error {
 	if err := m.store.Initialize(ctx); err != nil {

--- a/features/rbac/manager_hierarchy_test.go
+++ b/features/rbac/manager_hierarchy_test.go
@@ -5,6 +5,7 @@ package rbac
 import (
 	"context"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -29,6 +30,11 @@ func TestManager_CreateRoleWithParent(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	// Initialize manager
@@ -166,6 +172,11 @@ func TestManager_GetRoleHierarchy(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	// Initialize and set up test hierarchy
@@ -256,6 +267,11 @@ func TestManager_SetAndRemoveRoleParent(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -326,6 +342,11 @@ func TestManager_ComputeRolePermissions(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -395,6 +416,11 @@ func TestManager_ValidateHierarchyOperation(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)

--- a/features/rbac/manager_storage_test.go
+++ b/features/rbac/manager_storage_test.go
@@ -14,7 +14,6 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
-	business "github.com/cfgis/cfgms/pkg/storage/interfaces/business"
 	_ "github.com/cfgis/cfgms/pkg/storage/providers/database"
 	_ "github.com/cfgis/cfgms/pkg/storage/providers/flatfile"
 	_ "github.com/cfgis/cfgms/pkg/storage/providers/sqlite"
@@ -22,21 +21,14 @@ import (
 
 // TestNewManagerWithStorage tests the new constructor that accepts storage interfaces
 func TestNewManagerWithStorage(t *testing.T) {
+	// The database provider is exercised by its own provider-specific tests
+	// under pkg/storage/providers/database/. This test covers the OSS
+	// flatfile+sqlite composite path that the controller actually uses.
 	tests := []struct {
 		name         string
 		setupStorage func(t *testing.T) (*interfaces.StorageManager, error)
 		wantErr      bool
 	}{
-		{
-			name: "with database storage provider (if configured)",
-			setupStorage: func(t *testing.T) (*interfaces.StorageManager, error) {
-				// Skip database tests for now due to configuration complexity
-				// In production, database provider will be properly configured
-				t.Skip("database provider requires complex configuration - skipping for basic test")
-				return nil, nil
-			},
-			wantErr: false,
-		},
 		{
 			name: "with flatfile+sqlite storage provider",
 			setupStorage: func(t *testing.T) (*interfaces.StorageManager, error) {
@@ -156,34 +148,13 @@ func TestNewManagerWithStorage(t *testing.T) {
 // deliberately removed to eliminate package-level storage mechanisms
 // All RBAC managers now require explicit storage configuration via NewManagerWithStorage()
 
-// TestNewManagerWithStorage_NilStorageHandling tests error handling for nil storage interfaces
+// TestNewManagerWithStorage_NilStorageHandling verifies that NewManagerWithStorage
+// panics when any required store is nil. All three stores (audit, clientTenant,
+// rbac) are required — passing nil for any of them is a programmer error.
 func TestNewManagerWithStorage_NilStorageHandling(t *testing.T) {
-	tests := []struct {
-		name              string
-		auditStore        business.AuditStore
-		clientTenantStore business.ClientTenantStore
-		expectPanic       bool
-	}{
-		{
-			name:              "nil audit store",
-			auditStore:        nil,
-			clientTenantStore: nil,
-			expectPanic:       true,
-		},
-	}
-
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			if tt.expectPanic {
-				assert.Panics(t, func() {
-					NewManagerWithStorage(tt.auditStore, tt.clientTenantStore, nil)
-				})
-			} else {
-				manager := NewManagerWithStorage(tt.auditStore, tt.clientTenantStore, nil)
-				assert.NotNil(t, manager)
-			}
-		})
-	}
+	assert.Panics(t, func() {
+		NewManagerWithStorage(nil, nil, nil)
+	}, "NewManagerWithStorage must panic when all stores are nil")
 }
 
 // TestManagerWithStorage_TenantIsolation tests that tenant isolation works correctly with pluggable storage

--- a/features/rbac/manager_storage_test.go
+++ b/features/rbac/manager_storage_test.go
@@ -5,6 +5,7 @@ package rbac
 import (
 	"context"
 	"testing"
+	"time"
 
 	"github.com/cfgis/cfgms/api/proto/common"
 	"github.com/cfgis/cfgms/pkg/storage/interfaces"
@@ -63,6 +64,11 @@ func TestNewManagerWithStorage(t *testing.T) {
 				storageManager.GetClientTenantStore(),
 				storageManager.GetRBACStore(),
 			)
+			t.Cleanup(func() {
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				_ = manager.Close(ctx)
+			})
 			require.NotNil(t, manager)
 
 			// Verify manager initializes correctly with pluggable storage
@@ -193,6 +199,11 @@ func TestManagerWithStorage_TenantIsolation(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	require.NotNil(t, manager)
 
 	ctx := context.Background()
@@ -262,6 +273,11 @@ func TestManagerWithStorage_AuditTrail(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	require.NotNil(t, manager)
 
 	ctx := context.Background()

--- a/features/rbac/manager_test.go
+++ b/features/rbac/manager_test.go
@@ -5,6 +5,7 @@ package rbac
 import (
 	"context"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -29,6 +30,11 @@ func TestManager_Initialize(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -69,6 +75,11 @@ func TestManager_CreateTenantDefaultRoles(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -108,6 +119,11 @@ func TestManager_SubjectManagement(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -171,6 +187,11 @@ func TestManager_RoleAssignment(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -235,6 +256,11 @@ func TestManager_PermissionChecking(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -321,6 +347,11 @@ func TestManager_SystemAdminPermissions(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -386,6 +417,11 @@ func TestManager_CreateStewardSubject(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)
@@ -444,6 +480,11 @@ func TestManager_InactiveSubjectPermissions(t *testing.T) {
 		storageManager.GetClientTenantStore(),
 		storageManager.GetRBACStore(),
 	)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_ = manager.Close(ctx)
+	})
 	ctx := context.Background()
 
 	err = manager.Initialize(ctx)

